fml/message_loop_task_queues.cc

```diff
@@ -140,9 +164,136 @@ void MessageLoopTaskQueues::Swap(TaskQueueId primary, TaskQueueId secondary)
 void MessageLoopTaskQueues::SetWakeable(TaskQueueId queue_id,
                                         fml::Wakeable* wakeable) {
   std::scoped_lock lock(GetMutex(queue_id, MutexType::kWakeables));
+  FML_CHECK(!wakeables_[queue_id]) << "Wakeable can only be set once.";
   wakeables_[queue_id] = wakeable;
 }
 
+bool MessageLoopTaskQueues::Merge(TaskQueueId owner, TaskQueueId subsumed) {
+  // task_observers locks
+  std::mutex& o1 = GetMutex(owner, MutexType::kObservers);
+  std::mutex& o2 = GetMutex(subsumed, MutexType::kObservers);
+
+  // delayed_tasks locks
+  std::mutex& t1 = GetMutex(owner, MutexType::kTasks);
+  std::mutex& t2 = GetMutex(subsumed, MutexType::kTasks);
+
+  std::scoped_lock(o1, o2, t1, t2);
+
+  if (owner == subsumed) {
+    return true;
+  }
+
+  if (owner_to_subsumed_[owner] == subsumed) {
+    return true;
+  }
+
+  std::vector<size_t> owner_subsumed_keys = {
+      owner_to_subsumed_[owner], owner_to_subsumed_[subsumed],
+      subsumed_to_owner_[owner], subsumed_to_owner_[subsumed]};
+
+  for (auto key : owner_subsumed_keys) {
+    if (key != _kUnmerged) {
+      return false;
+    }
+  }
+
+  owner_to_subsumed_[owner] = subsumed;
+  subsumed_to_owner_[subsumed] = owner;
+
+  if (HasPendingTasksUnlocked(owner)) {
+    WakeUp(owner, GetNextWakeTimeUnlocked(owner));
+  }
+
+  return true;
+}
+
+bool MessageLoopTaskQueues::Unmerge(TaskQueueId owner) {
+  // task_observers locks
+  std::mutex& o = GetMutex(owner, MutexType::kObservers);
+
+  // delayed_tasks locks
+  std::mutex& t = GetMutex(owner, MutexType::kTasks);
+
+  std::scoped_lock(o, t);
```

Comment:
This does not actually perform the lock. Please use a variable here.

Reply:
Done